Adding support for S3-compatible Storage for Repository part

pedro-stanaka opened the pull request:
This PR is related to issue #228. I tried to keep it simple and contained, but there were too many places where it the local filesystem was taken as a guarantee. So there was a lot of refactoring involved.

I tried to follow the conventions you used in the code, but if I failed to respect some convention please let me know and I will fix it. 

Apart from the strong test suite of yours, I also tested the application inside a private hosted server with the ansible setup and it worked well for both the local and for the new S3 storage. The only problem that I could find is when migrating from local to S3, but I think in a later PR someone (or me) can provide a proper command to migrate files from local to S3 storage.

Once again, if you see any problems on the PR, let me know and I will fix in a manageable time, I spend almost a month working on this, so I would like to see it merged :smile_cat: . 

Thanks for the awesome tool you guys are providing for the community.

pedro-stanaka commented on src/Service/ReadmeExtractor.php:
```diff
@@ -51,10 +52,13 @@ public function extractReadme(Package $package, Dist $dist): void
     private function loadREADME(Dist $dist): ?string
     {
-        $zipFilename = $this->distStorage->filename($dist);
+        $tmpLocalFilename = $this->createTemporaryZipFile($dist);
```
This here I would like some comment on it. If anyone has a better idea on how to resolve this problem of reading a zip file from stream, would be nice. I saw this lib here: https://github.com/maennchen/ZipStream-PHP, but dunno if it is worth the effort to integrate with it. Also, there is always the TechDebt that comes with the lib itself.

akondas replied:
I don't think it will make much difference, and we still have to go through the entire archive, unless it's actually smart enough that it can only get the index of the files and then the readme itself. Either way, it can stay as it is at the moment. This will only happen once for a new package.

pedro-stanaka replied:
Okay, I am resolving this one. This library is much more focused on streaming zip responses, so I don't think it will provide this kind of control over reading one ZIP file.

pedro-stanaka commented on ansible/roles/app.install/tasks/main.yml:
```diff
@@ -22,4 +22,10 @@
+  with_items:
+    - "{{ app_releases_dir }}"
+    - "{{ app_shared_dir }}/var/security-advisories"
+    - "{{ app_shared_dir }}/var/log"
+    - "{{ app_shared_dir }}/var/cache"
+    - "{{ app_shared_dir }}/var/proxy"
```
Without these folders, my installation on Ubuntu 18.04 was failing.

pedro-stanaka reviewed (commented):
Last few comments from my side.

akondas commented on src/Service/Organization/PackageManager.php:
```diff
@@ -4,26 +4,31 @@
 use Munus\Control\Option;
-use Symfony\Component\Filesystem\Filesystem;
+use function serialize;
+use function unserialize;
```
Please use `global_namespace_import` cs fixer (you can add it to `.php_cd.dist`), but do not import this global functions.   We would like to fully qualifies global classes/functions/constants.

pedro-stanaka replied:
Did that, sorry for this one. It was autofix from my IDE. I added the configuration to the `.php_cd.dist` file so the next person won't do the same :+1:

akondas replied:
What IDE do you use if I can know?

pedro-stanaka replied:
I am using [PHPStorm](https://www.jetbrains.com/phpstorm/) with [EA Extended Inspections](https://plugins.jetbrains.com/plugin/7622-php-inspections-ea-extended-).

akondas replied:
Thanks for info :+1:

akondas reviewed (commented):
Ffirst of all, many thanks for the effort :clap: :beers: .
We are happy to merge these changes.

But first, please make these minor corrections. Then we will test these changes on our side.

If everything works, we will merge next week :tada:

akondas commented on src/Service/Dist/Storage/StorageImpl.php:
```diff
@@ -0,0 +1,100 @@
+use function sprintf;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+
+final class StorageImpl implements Storage
```
Ok, we will do it separately, no rush, so ignore my comment below.

pedro-stanaka replied:
Nice, as soon as this gets merged I commit myself on doing a PR to clean this up :crossed_fingers:

akondas commented on src/Service/Dist/Storage/StorageImpl.php:
```diff
@@ -0,0 +1,100 @@
+use function sprintf;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+
+final class StorageImpl implements Storage
```
above is a comment, for now it will remain as is :+1:

pedro-stanaka commented on .env.docker:
```diff
@@ -42,8 +42,8 @@ DATABASE_URL=postgresql://main:main@database:5432/main?serverVersion=11&charset=
-# MESSENGER_TRANSPORT_DSN=amqp://guest:guest@localhost:5672/%2f/messages
-# MESSENGER_TRANSPORT_DSN=doctrine://default
+#MESSENGER_TRANSPORT_DSN=amqp://guest:guest@localhost:5672/%2f/messages
+ MESSENGER_TRANSPORT_DSN=doctrine://default
```
I was trying to figure out, why TH my tests suddenly started to fail locally. Now I know why... :facepalm: 

Commented again.

pedro-stanaka commented on src/Service/AmazonWebServices/S3AdapterFactory.php:
```diff
@@ -0,0 +1,53 @@
+
+        if ($this->isOpaqueAuth) {
+            if ($key === null) {
+                throw new \InvalidArgumentException('Must pass AWS key when authentication is opaque');
```
Added tests with some data provider to test different scenarios, thx for seeing this.

akondas replied:
Thanks for test :+1:

akondas commented:
I must admit that you create very elegant commit descriptions. I appreciate :clap:

akondas commented:
After initial testing I found two small bugs, I push a fix.
However, the scanning of packages still does not work. Look like `Storage::filename` return wrong path, before this PR it looks like this:
```php
public function filename(Dist $dist): string
    {
        return sprintf(
            '%s/%s/dist/%s/%s_%s.%s',
            $this->distsDir,
            $dist->repo(),
            $dist->package(),
            $dist->version(),
            $dist->ref(),
            $dist->format()
        );
    }
```

As you can see, full relative path from content root is returned.

akondas commented on src/Service/Organization/PackageManager.php:
```diff
@@ -67,9 +67,11 @@ public function saveProvider(array $json, string $organizationAlias, string $pac
+        }
 
-        $this->repoFilesystem->write($filepath, \serialize($json));
+        $this->repoFilesystem->put($filepath, \serialize($json));
```
write will throw exception if file exist :wink:

pedro-stanaka replied:
Now I got this, you actually provided the fix. Thx.

pedro-stanaka commented:
> After initial testing I found two small bugs, I push a fix.
> However, the scanning of packages still does not work. Look like `Storage::filename` return wrong path, before this PR it looks like this:
> 
> ```
> public function filename(Dist $dist): string
>     {
>         return sprintf(
>             '%s/%s/dist/%s/%s_%s.%s',
>             $this->distsDir,
>             $dist->repo(),
>             $dist->package(),
>             $dist->version(),
>             $dist->ref(),
>             $dist->format()
>         );
>     }
> ```
> 
> As you can see, full relative path from content root is returned.

Now we are passing the `$this->distDir` as base for the filesystem so we shouldn't need this part anymore. I will take a look this afternoon and add more tests if needed. Same for the  `Filesystem::put()` problem.

pedro-stanaka commented:
Regarding: https://github.com/repman-io/repman/pull/332#issuecomment-739747234

How did you manage to reproduce the bug? Can you help me by pointing a class or giving me the exception stack trace so I can try to write a test case?

akondas commented:
> Regarding: [#332 (comment)](https://github.com/repman-io/repman/pull/332#issuecomment-739747234)
> 
> How did you manage to reproduce the bug? Can you help me by pointing a class or giving me the exception stack trace so I can try to write a test case?

I have a package added earlier and I run security scanner for it. I think there will be the same problem for the newly added package. Security scanner probably wasn't tested externally, only as a separate service and that's why this problem.

pedro-stanaka commented:
Hi @akondas sorry for the immense delay in my response. I tried to test locally and now I am getting one error while doing the security check, is this by any chance the same error as you are getting? 

![image](https://user-images.githubusercontent.com/2609731/102588435-bb702000-410d-11eb-9d32-ac3afd866f84.png)

If so, can you tell me why it does not get logged neither reported to sentry?

pedro-stanaka commented:
AH! I think I got it. I will fix it ASAP. 😃 I found one place where I didnt change how we are reading the Zip files. But still, I was hoping this exception was going to be reported to Sentry.

akondas commented:
> AH! I think I got it. I will fix it ASAP.  I found one place where I didnt change how we are reading the Zip files. But still, I was hoping this exception was going to be reported to Sentry.

Yes, in prod env it will be reported to sentry :wink:

pedro-stanaka commented:
Okay, I think I managed to solve this SecurityScanner problem. I will use the application a bit more locally to be sure I dont find any other problems, but I guess it is fine.

pedro-stanaka commented:
@akondas now I only have a problem with this conflict, if you could help me resolve it would be nice. I tried to resolve locally by doing a rebase, but had to run `composer update --lock -W` which upgraded many dependencies and ended up totally breaking the tests (and I suppose the application itself).

akondas commented:
Yes, I can help you with this. I will correct it immediately when testing this branch

akondas commented on tests/Unit/Service/AmazonWebServices/S3AdapterFactoryTest.php:
```diff
@@ -0,0 +1,93 @@
+ * THIS FILE IS A PART OF CRM PROJECT
+ * CRM PROJECT IS PROPERTY OF Legal One GmbH.
+ *
+ * @copyright Copyright (c) 2020 Legal One GmbH (http://www.legal.one)
```
was this file copied or maybe it is unnecessary?

pedro-stanaka replied:
HAHA, this is my IDE configuration, disabled the configuration for this project.

akondas replied:
ok, thanks for fix, rebased with master, so if everything will be working we will merge this :wink:

karniv00l commented on tests/Doubles/FakeDownloader.php:
```diff
@@ -18,8 +18,9 @@ final class FakeDownloader implements Downloader
-    public function __construct(string $basePath = __DIR__.'/../Resources')
+    public function __construct(?string $basePath = null)
     {
+        $basePath = $basePath ?? dirname(__DIR__, 1).DIRECTORY_SEPARATOR.'Resources';
```
what's the reason for level 1 in dirname and use of DIRECTORY_SEPARATOR?

pedro-stanaka replied:
Was just to avoid the relative paths with `../` on it. I had problems when using those in conjunction with the composer repositories, specially the paths. Regarding the `DIRECTORY_SEPARATOR`, I try to use it whenever I can for portability. If you guys prefer just `/` I can replace it as well. 

Now running the suite again with the old code I see that I don't get any errors :man_shrugging: . So I can easily revert it, just say the word. 😃

karniv00l replied:
It's ok, it's just that `1` is a default level for the `dirname` function and we don't really support windows in other places 😅

pedro-stanaka replied:
OMG, I didn't even see that (default value). I will go back to the old code to avoid confusion.

pedro-stanaka replied:
There you go! Thanks for catching such fine grained detail :+1:

karniv00l replied:
Cheers, good job overall 🍻

akondas commented:
We tested everything (for the local filesystem) and it looks like it works, we still have s3 configuration to check. Due to the fact that holidays are coming, we will probably merge this branch in the first week of January. Forgive me for taking so long, but I would like to supervise this myself. Thanks for the work and Merry Christmas.

pedro-stanaka commented:
Wishing you a late Merry Christmas and all good for the new year!!! I
actually deployed this branch to production at my company to do a test run
before the holidays, if I face some problem I will also provide a patch
here. Until next year! 🖖
----
Pedro Tanaka


On Wed, Dec 23, 2020 at 3:18 PM Arkadiusz Kondas <notifications@github.com>
wrote:

> We tested everything (for the local filesystem) and it looks like it
> works, we still have s3 configuration to check. Due to the fact that
> holidays are coming, we will probably merge this branch in the first week
> of January. Forgive me for taking so long, but I would like to supervise
> this myself. Thanks for the work and Merry Christmas.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/repman-io/repman/pull/332#issuecomment-750319653>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAT5EQ2BAVFOKKQ6DPGZFDDSWH3Z5ANCNFSM4UNHXM3A>
> .
>

akondas commented:
Hello @pedro-stanaka, I hope all is well in the new year. Tests showed that managing the private packages works flawlessly, but we had a small problem with the proxy on s3. I am attaching an error dump:

![Screenshot from 2021-01-04 11-54-54](https://user-images.githubusercontent.com/8239917/103529500-13d14c80-4e86-11eb-898c-cbdaf3caf210.png)

pedro-stanaka commented:
Okay, I will fix this ASAP. I already know the problem, the file is on remote location and [fstat requires local file system file](https://www.php.net/manual/en/function.fstat.php#refsect1-function.fstat-notes). 

@akondas would it be okay if we pass the file size inside the constructor for `src/Service/Proxy/Metadata.php`?

akondas commented:
I guess that should be ok 👍. Too bad I thought fstat worked for stream data as well.

pedro-stanaka commented:
@akondas I think it should be fine now. I introduced a class for the Dist File for the proxy, because there we really only had the stream, so I added class to have information about the file size we are proxying. Hope solution is fine enough.

akondas commented on src/Service/Proxy/DistFile.php:
```diff
@@ -2,7 +2,7 @@
+ * CRM PROJECT IS PROPERTY OF Legal One GmbH.
  *
  * @copyright Copyright (c) 2020 Legal One GmbH (http://www.legal.one)
  */
```
:scissors: :wink:

akondas commented:
Everything seems to be working properly now. Thanks for a great job. Please delete this one comment and we are ready to merge it :tada:

akondas reviewed (changes requested):
One more thing to fix: `repman:metadata:clear-cache` cli command. This is last place where local files are used :wink:.

pedro-stanaka commented:
I will work on solution for the cli command. The symfony filesystem component is a really nice helper, but wont work with the new Flysystem adapter. I will have devise a "by-hand" solution, by iterating the contents of the distDir. Luckily this is cronjob, so no problem taking a bit more of time.

pedro-stanaka commented:
@akondas I managed to find a solution for the file finding. I took " inspiration" from symfony filesystem lib. Please take a look and see if this is adequate. 👍

akondas reviewed (approved):
LGTM :+1: :clap:

akondas commented:
@pedro-stanaka Thank you for all your hard work on this functionality :pray: :clap: :medal_military:
